Doppler.BigQueryMicroservice/Services/EmailSender.cs

```diff
@@ -0,0 +1,82 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Tavis.UriTemplates;
+using Doppler.BigQueryMicroservice.Utils;
+using Microsoft.Extensions.Logging;
+
+namespace Doppler.BigQueryMicroservice.Services
+{
+    public class EmailSender : IEmailSender
+    {
+        private readonly ILogger<EmailSender> _logger;
+        private readonly RelayEmailSenderConfiguration _config;
+        private IFlurlClient _flurlClient = null;
+
+        private IFlurlClient FlurlClient => EnsureAndGetFlurlClient();
+
+        public EmailSender(IOptions<RelayEmailSenderConfiguration> config, ILogger<EmailSender> logger)
+        {
+            _config = config.Value;
+            _logger = logger;
+        }
+
+        ///
+        public async Task<bool> SendWithTemplateAsync(string templateId, object templateModel, IEnumerable<string> to, IEnumerable<string> cc = null,
+                                                IEnumerable<string> bcc = null, string fromName = null, string fromAddress = null, IEnumerable<Attachment> attachments = null,
+                                                CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                var recipients = (
+                    from emailAddress in to.EmptyIfNull()
+                    select new { email = emailAddress, type = "to" }).Union(
+                    from emailAddress in cc.EmptyIfNull() select new { email = emailAddress, type = "cc" }).Union(
+                    from emailAddress in bcc.EmptyIfNull() select new { email = emailAddress, type = "bcc" }).ToArray();
+
+                await FlurlClient
+                    .Request(new UriTemplate(_config.SendTemplateUrlTemplate)
+                        .AddParameter("accountId", _config.AccountId)
+                        .AddParameter("accountName", _config.AccountName)
+                        .AddParameter("username", _config.Username)
+                        .AddParameter("templateId", templateId)
+                        .Resolve())
+                    .PostJsonAsync(new
+                    {
+                        from_name = fromName ?? _config.FromName,
+                        from_email = fromAddress ?? _config.FromAddress,
+                        recipients = recipients,
+                        attachments = attachments?.Select(x => new
+                        {
+                            content_type = x.ContentType.ToString(),
+                            base64_content = Convert.ToBase64String(x.Content),
+                            filename = x.Filename
+                        }),
+                        model = templateModel
+                    }, cancellationToken);
+
+                return true;
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, $"Error in method SendWithTemplateAsync, Service  {nameof(EmailSender)}");
+                throw;
```

Comment:
Por ahora está bien, pero cuidado con esta excepción cuando esto sea parte de la operación normal.

Osea, si funciona el PUT, no debería fallar porque no se envía el email.

Reply:
suggest accepted and applied